lib/filestorage.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const common = require('metarhia-common');
+const utils = require('./utils.js');
+
+class FileStorage {
+
+  // Create new FileStorage
+  //   options - <object>
+  //     dir - <string>, data storage directory
+  //     minCompressSize - <number>, minimal file size to be compressed
+  constructor(options) {
+    this.dir = path.resolve(options.dir);
+    this.minCompressSize = options.minCompressSize;
+    this.compressionList = [];
+    this.rmList = [];
+  }
+
+  // Write file to storage
+  //   id - <Object>, common.Uint64, id of file
+  //   data - <string> | <Buffer> | <Uint8Array>, data to be written
+  //   opts - <Object>
+  //     checksum - <string>, checksum type
+  //     dedupHash - <string>, second checksum type
+  //   cb - <Function>, callback
+  //     err - <Error>
+  //     checksum - <string>, data checksum
+  //     dedupHash - <string>, second data checksum
+  //     originalSize - <number>, data size
+  // Throws: <TypeError>, if `opts.checksum` or `opts.dedupHash` is incorrect
+  write(id, data, opts = {}, cb) {
+    const checksum = utils.computeHash(data, opts.checksum || 'CRC32');
+    const dedupHash = utils.computeHash(data, opts.dedupHash || 'SHA256');
+    const originalSize = data.length;
+
+    const file = utils.getFilepath(this.dir, common.idToPath(id));
+    utils.mkdirp(path.dirname(file), err => {
+      if (err) {
+        cb(err);
+      } else {
+        fs.writeFile(file, data, err => {
+          cb(err, checksum, dedupHash, originalSize);
+        });
+      }
+    });
+  }
+
+  // Read file from storage
+  //   id - <Object>, common.Uint64, id of file
+  //   cb - <Function>, callback
+  //     err - <Error>
+  //     data - <Buffer>
+  read(id, cb) {
+    const file = utils.getFilepath(this.dir, common.idToPath(id));
+    fs.readFile(file, cb);
```

Comment:
Shouldn't you also check if `filename` + 'z' (compressed) is present?

Reply:
The original file is replaced with compressed one and since database knows it was compressed I don't think we should return original file instead of compressed. In case if we are going to store both files then we need to add a separate method like `readOriginal`.

Reply:
How will database pass you the correct path? (This method accepts only `id` not filepath or `id + z` as that will be incorrect ID).